Logo control modification

acalcutt opened the pull request:
I've created a new logic for showing the Maplibre logo, like mentioned in ( https://github.com/maplibre/maplibre-gl-js/issues/121 ). I think this gives the maplibre project a bit more visibility by default, while still allowing the logo to be turned off.  

Right now the logo only shows if the TileJSON source specifies the 'maplibreLogo' option, but I think for the most part this means it never gets shown. What I have done is moved the maplibreLogo option from inside the TileJSON to being a map option. 

```
var map = new maplibregl.Map({
    container: 'map', // container id
	style: 'https://demotiles.maplibre.org/style.json', 
    maplibreLogo: true,
    logoPosition: "bottom-left"
});
```

If maplibreLogo is true or unspecified, the logo will show. If maplibreLogo is set to false, it will not show.  The logoPosition option can still be used to control the position (it will be in bottom-left if not define) like before.

Edit: here is a demo page ( https://stackblitz.com/edit/web-platform-f1k3ah?file=index.html )
![logo_off](https://user-images.githubusercontent.com/3792408/149675298-cb1c9260-63e8-4ae2-b051-fb34e2361f62.png)
![logo_on](https://user-images.githubusercontent.com/3792408/149675299-74a584c8-dde0-4051-a1a4-3cd95c83951d.png)

I've also updated the tests for the new behavior of logo control. 

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [X] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

HarelM commented:
Can you fix the lint errors? `npm run lint -- --fix` can help...

HarelM commented:
This is a breaking change in terms of how the UI will look for someone who updates the version.
I tend to say that another flag is a bit redundant.
I do think that it should not be part of the tileJson as you wrote.
I think a simpler approach would be to check if `logoPosition` is set in the options - if it exists then place the logo at that position, otherwise don't place the logo at all.
It might be breaking change as well when thinking about it... IDK...

acalcutt commented:
The lint issues should be fixed.

The logoPosition is an existing option that sets what corner the logo displays I don't think it would be useful to set whether the logo shows or not since it is optional and doesn't need to be specified at all (it it is not set it defaults to ControlPosition which is bottom-left)

Really the "maplibreLogo" option I added is to give users the option to turn the logo off. if it is not included the default is to show the logo.

acalcutt commented:
If you wanted to make it less of a change to someone who upgrades, it could be set so the logo defaults off unless the user specifies "maplibreLogo: true" but I though that kind of defeated the purpose a bit

Testing it myself I saw how it just being there suddenly could be an issue. for example, when I added it to my own site with the defaults, it overlapped some controls I had put in that corner and I had to place it elsewhere with logoPosition
![logo_over_controls](https://user-images.githubusercontent.com/3792408/149680396-6173d82c-02ed-402e-a404-988dc90e73b9.png)

acalcutt commented:
I've updated this a bit more so it works more similarly to attribution control, where it can be added with addControl
`map.addControl(new maplibregl.LogoControl(), 'bottom-right');`

My thinking was if it was decided that this shouldn't be show by default, the options and code to show the logo in map.ts could probably just be removed. Then the logo would only exist if someone specifically added it with addControl .

HarelM commented:
We can make this on by default, but we'll need a "breaking change" version, which we just did with version 2, so I'm not super exited to create a new breaking change version soon...
Never-the-less thinking about the logo as a control seems like a good direction for me.
We can make it on by default later on when we have a version with breaking changes anyway...

acalcutt commented:
I've been testing it out a bit, and it is defiantly a lot more flexible as a control. it is easier to place it exactly where you want in relation to other controls. I do think the built in default one is useful for branding (getting people to know maplibre exists), but at least you can turn it off an customize it.

I also added in a 'compact' option similar to attribution control, so when you manually add it you can force it to always be compact or always be full size. if you leave it unset it changes at the same time attribution control changes, which looks better when they change in unison.

Heres an example. 
( https://stackblitz.com/edit/web-platform-xfhypz?file=index.html )
the logo in the bottom-left is the default one, which can be turned off by setting 'maplibreLogo' to false. In the default logo compact is undefined, so if it is above 640px the full version shows, if it is below 640 the compact version shows.

the logo in the top-left is the one I added with addControl. that logo has the compact option set to false, so it always shows full size. the compact option could be set to 'true' to make it always be compact.
![compact](https://user-images.githubusercontent.com/3792408/149961917-80357dd6-179a-4e44-b66e-8b4a9ab3c84f.png)

acalcutt commented:
This should be updated to not display by default now.

There are two ways to show it now, 

1.) Setting maplibreLogo map option to true
```
var map = new maplibregl.Map({
	container: 'map', // container id
	style: 'https://demotiles.maplibre.org/style.json', 
	maplibreLogo: true,
});

```
2.) Adding it manually with add control
```
map.addControl(new maplibregl.LogoControl({
        compact: false
    }),
    'top-left'
);
```

HarelM commented:
@acalcutt Can you please fix conflicts?
@wipfli @birkskyum can you guys review this as well?

acalcutt commented:
This should be up to date with main now

wipfli commented:
Thanks for this pull request @acalcutt. I think it is best if the logo does not show up automatically if someone updates to a new version of MapLibre GL JS. Before, the logo was not shown by default so the new default should be to not show the logo either.

> it could be set so the logo defaults off unless the user specifies "maplibreLogo: true" but I though that kind of defeated the purpose a bit

I'm sorry if I did not catch this point earlier in the review. Is it an option for you to not show the logo by default?

wipfli commented:
I need to check that the default is indeed to not show the logo. It would be cool if stackblitzes could use the latest version of a pull request automatically somehow...

HarelM commented:
We should consider publishing the compiled file to the pull request somehow, it would be super useful.

wipfli commented:
#328 might be a nice starting point for automatic `maplibre-gl*` file deployment.

acalcutt commented:
I can confirm I updated it to not show my default. 

There are two ways to show it now,

1.) Setting maplibreLogo map option to true

var map = new maplibregl.Map({
	container: 'map', // container id
	style: 'https://demotiles.maplibre.org/style.json', 
	maplibreLogo: true,
});

2.) Adding it manually with add control

map.addControl(new maplibregl.LogoControl({
        compact: false
    }),
    'top-left'
);

wipfli commented:
```js
new maplibregl.Map({
    container: 'map',
    zoom: 12.5,
    center: [-77.01866, 38.888],
    style: 'https://demotiles.maplibre.org/style.json',
    hash: true,
    maplibreLogo: true,
    logoPosition: 'top-left',
});
```
still shows the logo in the bottom-left corner. Is this intended?

acalcutt commented:
It should obey LogoPosition. ( https://github.com/maplibre/maplibre-gl-js/blob/69dc1ebbf918a26bb4df9e7c33178878ea967e99/src/ui/map.ts#L470 )

 I just updated this test page and it seems to work fine for me
https://stackblitz.com/edit/web-platform-mcez6s?file=index.html

wipfli commented:
My bad, I just did not clear the browser cache :S

wipfli commented on src/source/load_tilejson.ts:
```diff
@@ -16,7 +16,7 @@ export default function(options: any, requestManager: RequestManager, callback:
                 // explicit source options take precedence over TileJSON
                 extend(tileJSON, options),
-                ['tiles', 'minzoom', 'maxzoom', 'attribution', 'maplibreLogo', 'bounds', 'scheme', 'tileSize', 'encoding']
+                ['tiles', 'minzoom', 'maxzoom', 'attribution', 'bounds', 'scheme', 'tileSize', 'encoding']
```
What does this change do?

acalcutt replied:
The logo control used to only get displayed if a TileJSON source included 'maplibreLogo' . My changes removes it from TileJSON and makes it a map option instead, so getting this value from the TileJSON source is no longer needed

From the first post
> Right now the logo only shows if the TileJSON source specifies the 'maplibreLogo' option, but I think for the most part this means it never gets shown. What I have done is moved the maplibreLogo option from inside the TileJSON to being a map option.

wipfli replied:
Thanks for the clarification @acalcutt. Do you know where the TileJSON stuff is documented?

wipfli replied:
I'm asking because this seems like a breaking change in terms how we handle TileJSON.

acalcutt replied:
I looked a bit, but I don't even see logo control in the examples. Before it was more a built in thing for mapbox sources, so maybe it wasn't documented before ( https://github.com/mapbox/mapbox-gl-js/issues/4242 )?

wipfli commented on src/ui/control/logo_control.ts:
```diff
@@ -1,28 +1,43 @@
+        this.options = options;
+
+        bindAll([
+            '_updateLogo',
```
Is `_updateLogo` still necessary? You have removed the method from the class...

acalcutt replied:
Fixed, removed this

wipfli commented on src/ui/map.ts:
```diff
@@ -198,6 +200,7 @@ const defaultOptions = {
  * @param {number} [options.clickTolerance=3] The max number of pixels a user can shift the mouse pointer during a click for it to be considered a valid click (as opposed to a mouse drag).
  * @param {boolean} [options.attributionControl=true] If `true`, an {@link AttributionControl} will be added to the map.
  * @param {string | Array<string>} [options.customAttribution] String or strings to show in an {@link AttributionControl}. Only applicable if `options.attributionControl` is `true`.
+ * @param {boolean} [options.maplibreLogo=false] If `true`, the maplibre logo will be shown.
```
```suggestion
 * @param {boolean} [options.maplibreLogo=false] If `true`, the MapLibre logo will be shown.
```

acalcutt replied:
Fixed, updated the text

wipfli commented on src/util/test/util.ts:
```diff
@@ -9,6 +9,7 @@ export function createMap(options?, callback?) {
         container,
         interactive: false,
         attributionControl: false,
+        maplibreLogo: false,
```
Do the tests fail without this new line?

wipfli replied:
I am asking because I think the default value is now `false` anyway...

acalcutt replied:
Its here more for consistency in testing. With this here, the tests are consistent whether the maplibreLogo is default to true or false in map.ts. So in the future, if we decide to make this default to true, the tests should continue to work.

Before the logo control always showed, so some of the tests expected that control to exist. When I made LogoControl optional and false by default, there were map test that failed because they expected one more control to exists. I updated the tests to not expect LogoControl by default.

I followed the example of attributionControl, which is similarly optional in map.ts. That control also has a default set in this file on line 11 ( attributionControl: false ), so it made sense to do this one the same way.

wipfli replied:
I understand that it makes sense to have it consistent.

> So in the future, if we decide to make this default to true, the tests should continue to work.

But actually we would want the opposite, right? If someone changes the default, then the tests should fail. I would say let's remove this `maplibreLogo: false`.

acalcutt replied:
It was more for the map.ts tests. Because the failure in the map.ts tests was hard to understand at first, when the similar attibution control was working whether it was defaulted to true or false. it took a lot to figure out, that was because attibution control had a default in this util.ts file, and that was why attribution controls tests were consistent and maplibreLogo were not.

The actual logo control tests don't actually use this since they have their own CreateMap function. If the default was changed in map.ts, there would be a failure marked in this test ( https://github.com/maplibre/maplibre-gl-js/blob/47394d1ae0ed1dd75f10fc20574832bfd642962a/src/ui/control/logo_control.test.ts#L23-L32 ) . You can see in that it uses the custom CreateMap function on line 3.

wipfli replied:
OK, thanks for the clarification!

wipfli commented on src/ui/control/logo_control.ts:
```diff
@@ -44,40 +56,19 @@ class LogoControl implements IControl {
             const anchor = containerChildren[0];
-            if (this._map.getCanvasContainer().offsetWidth < 250) {
-                anchor.classList.add('maplibregl-compact', 'mapboxgl-compact');
+            if (this._map.getCanvasContainer().offsetWidth <= 640 || this._compact) {
```
I think you @acalcutt and @astridx had discussions about the right numeric values for this in the past. Do you mind linking them for me? The new value seems reasonable to me but I can't remember how you got there.

acalcutt replied:
There wasn't much conversation on this, but I mentioned it here ( https://github.com/maplibre/maplibre-gl-js/pull/786#issuecomment-1015496885 )

I made this value the same as attribution control, because it looked better when they changed size together. So when you shrink a windows, both attribution control and logo control go to compact mode at the same time.

wipfli commented:
Thanks @acalcutt for your patience. There are some minor questions but overall I think this is good!

wipfli commented:
I forgot that we need a changelog entry. Can you make a separate pull request for this?
